types/src/jsonrpc/wrapped/batches.rs

```diff
@@ -180,32 +178,18 @@ impl<T> BatchesState<T> {
 	/// Injects a newly-received batch into the list. You must then call
 	/// [`next_event`](BatchesState::next_event) in order to process it.
 	pub fn inject(&mut self, request: jsonrpc::Request, user_param: T) {
+		let batch_id = self.vacant.pop().unwrap_or_else(|| self.batches.len());
 		let batch = batch::BatchState::from_request(request);
 
-		loop {
-			let id = self.next_batch_id;
-			self.next_batch_id = self.next_batch_id.wrapping_add(1);
-
-			// We shrink `self.batches` from time to time so that it doesn't grow too much.
-			if id % 256 == 0 {
-				self.batches.shrink_to_fit();
-				// TODO: self.batches.shrink_to(BATCHES_MIN_CAPACITY);
-				// ^ see https://github.com/rust-lang/rust/issues/56431
-			}
-
-			match self.batches.entry(id) {
-				Entry::Occupied(_) => continue,
-				Entry::Vacant(e) => {
-					e.insert((batch, user_param));
-					break;
-				}
-			}
-		}
+		self.batches.insert(batch_id, Some((batch, user_param)));
```

Comment:
Doesn't this shift all elements to the right? We know it is ok to clobber this index so could we not do better than this?

Reply:
This was me being clever, but this is actually a bug, it should either push if there are no vacant ids, or overwrite an existing `None` otherwise.

Reply:
Yup, this also fixed the failing integration test!